Add StyleSheet.flatten

mroswald opened the pull request:
adding StyleSheet.flatten

mroswald committed:
Add StyleSheet.flatten

mroswald commented:
rebased to current master

mroswald commented:
@lelandrichardson anything missing here?

lelandrichardson commented:
@mroswald thanks for the contribution! this isn't really doing any flattening though...

`StyleSheet.flatten` needs to allow arbitrarily nested arrays of styles.

```
flatten([foo, [bar, baz], bux])
```
is roughly equivalent to:
```
Object.assign({}, foo, bar, baz, bux])
```

mroswald committed:
improve StyleSheet.flatten to match the Api, add test

mroswald commented:
@lelandrichardson Ok, I fixed that and added the test cases specified in the React Native documentation: https://facebook.github.io/react-native/docs/stylesheet.html

lelandrichardson commented on src/api/StyleSheet.js:
```diff
@@ -5,6 +5,13 @@ const StyleSheet = {
   },
+  flatten(styles) {
+    if (Array.isArray(styles)) {
+      return Object.assign({}, ...styles);
```
quick tweak:

```
return Object.assign({}, ...styles.map(flatten))
```

Which should make this pass for nested style arrays.

lelandrichardson commented:
@mroswald thanks for the quick turnaround! I had one quick suggestion if you don't mind.

mroswald committed:
add support for nested arrays (recursive)

mroswald commented:
@lelandrichardson aye! :-)
